graphene_plugin.py

```diff
@@ -1,381 +1,690 @@
+
+    return _get_python_type_from_graphene_field_first_argument(
+        semanal, expression.args[0], covariant=covariant, nullable=not _is_required_kwarg_true(expression)
     )
 
 
-def get_metaclass_attribute_types(class_body: List[Statement],
-                                  ctx: ClassDefContext) -> Optional[List[Optional[TypeNodeInfo]]]:
-    # Weakly support interfaces.
-    interface_attributes: List[Optional[TypeNodeInfo]] = []
-
-    meta_classes = [
-        attribute for attribute in class_body if isinstance(attribute, ClassDef) and attribute.name == 'Meta'
-    ]
-    interfaces: List[AssignmentStmt] = []
-    if meta_classes:
-        meta_class_body = meta_classes[0].defs.body
-        interfaces = [
-            attribute for attribute in meta_class_body
-            if isinstance(attribute, AssignmentStmt) \
-            and attribute.lvalues[0].name == 'interfaces' # type: ignore[attr-defined]
-        ]
-
-    if not interfaces:
-        return interface_attributes
-
-    tuple_expr = interfaces[0].rvalue
-    if not isinstance(tuple_expr, TupleExpr):
-        not_a_tuple_error_message = '"interfaces" attribute in Meta class must be a tuple type'
-        if hasattr(tuple_expr, 'line'):  # This unknown node type can be used as context for an error
-            ctx.api.fail(not_a_tuple_error_message, tuple_expr)
+def _get_python_type_from_graphene_argument_instantiation(
+    semanal: SemanticAnalyzerPluginInterface, expression: CallExpr
+) -> Type:
+    """
+    Given an `Argument()` defintion, return the python type that will be passed to the resolver at runtime.
+
+    E.g. `Argument(List(NonNull(String)), required=True) -> builtins.list[builtins.str]`
+    """
+
+    assert isinstance(expression.callee, NameExpr) and isinstance(expression.callee.node, TypeInfo)
+
+    graphene_type: Optional[Expression] = None
+    if _type_is_a(expression.callee.node, GRAPHENE_UNMOUNTED_TYPE_NAME):
+        # What we actually have here is some graphene type being instantiated directly
+        # (e.g. `foo = String(required=True)` instead of `foo = Argument(String, required=True)`)
+        # In this case, the type of the arg is actually the thing being instantiated, not the first arg
+        if _type_is_a(expression.callee.node, GRAPHENE_STRUCTURE_NAME):
+            # This is a Graphene `Structure` (`List()`/`NonNull()`). Graphene `Structure`s wrap another
+            # type, so we actually want to treat the entire `CallExpr` as the thing to look at
+            # (e.g. `List(String)`) and not just the callee (e.g. `List`).
+            graphene_type = expression
         else:
-            ctx.api.fail(not_a_tuple_error_message, tuple_expr)
-        return interface_attributes
+            # This is just a regular `UnmountedType`
+            graphene_type = expression.callee
+    elif expression.args:
+        # We have an `Argument()` instantation. Get the type from the first argument
```

Comment:
This is probably a fair assumption, but should we make absolutely sure this is true? If I defined `bad_field = TotallyRandomClass('hi_mom')` in my `ObjectType` class, would it cause the plugin to crash mypy?

Reply:
Maybe it's out of the scope of this change but even if it doesn't crash mypy it's probably worth checking for and raising a useful error in the case above since it would be a runtime error.

Reply:
By the time we call this function, we've already checked that the `CallExpr` `expression` being passed in is either an `Argument(...)`  or `SomeScalar(...)`. So if you did `Field(String, bad_arg=TotallyRandomClass('hi_mom'))` our plugin would not crash. But, as you've pointed out, it would also not tell you that what you did is bad. I think that could be a useful feature but I'd rather punt on it! I think a feature like that would actually be great to open an issue for—then maybe somebody in the community could pick it up. Seems like it'd be a pretty good "first PR" feature.